### PR TITLE
Improved fab file and code base in general

### PR DESCRIPTION
**Pull request by ckcollab**
Resolves issue #999 

## Todo
- [x] Revert some of the line-length changes to be clearer

## Waiting on confirmation from
- [ ] @Raisins, received on _________
- [ ] @percyliang, received on _________

### CODE REVIEW

File: codalab/apps/analytics/views.py [hunk 1]
```diff
@@ -19,13 +19,21 @@ def analytics_detail(request):
...
-        users_joined_this_month = User.objects.filter(date_joined__month=date.month,
-                                                      date_joined__year=date.year).aggregate(Count('pk'))
-        monthly_total_users_joined[date.year][date.strftime("%B")] = users_joined_this_month['pk__count']
+        users_joined_this_month = User.objects.filter(
+            date_joined__month=date.month,
+            date_joined__year=date.year).aggregate(
+            Count('pk'))
+        monthly_total_users_joined[
+            date.year][
+            date.strftime("%B")] = users_joined_this_month['pk__count']
```
Comment (Raisins) [inline codalab/apps/analytics/views.py]: I know you are changing this so it is pep8, but i think 80 line count is a bit much it was easier to understand and read before.  In my linter I run local i have error `E501` off "line too long

https://github.com/jcrocholl/pep8/blob/master/pep8.py#L198

this also applies to the render change below.

File: codalab/apps/api/routers.py [hunk 2]
```diff
@@ -1,48 +1,104 @@
...
+                url(r'^bundles/create/$',
+                    views.BundleCreateApi.as_view(),
+                    name='api_bundle_create'),
+                url(r'^bundles/upload_url/$',
+                    views.BundleUploadApi.as_view(),
+                    name='api_bundle_upload_url'),
+                url(r'^bundles/(?P<uuid>[A-Za-z0-9]+)/$',
+                    views.BundleInfoApi.as_view(),
+                    name='api_bundle_info'),
+                )
```
Comment (Raisins) [inline codalab/apps/api/routers.py]: see comment above about readability

File: codalab/apps/web/management/commands/random_competitions.py [hunk 3]
```diff
@@ -1,106 +1,110 @@
...
+#                                                       datafile=f)
+#
+#                 CompetitionPhase.objects.create(competition=c,
+#                                                 phasenumber=i,
+#                                                 label="Phase %d" % i,
+#                                                 start_date=start_date,
+#                                                 max_submissions=4,
+#                                                 dataset=d)
+#
+#                 delta = datetime.timedelta(days=10)
```
Comment (Raisins) [inline codalab/apps/web/management/commands/random_competitions.py]: why is this commented out?

File: codalab/codalabtools/compute/worker.py [hunk 4]
```diff
@@ -135,16 +145,33 @@ def getThem(bundle_id, bundle_rel_path, bundles, depth):
...
+                        "command",
+                        "exitCode",
+                        "elapsedTime",
+                        "stdout",
+                        "stderr",
+                        "submitted-by",
+                        "submitted-at"):
                     if isinstance(v, str):
-                        getThem(v, join(bundle_rel_path, k), bundles, depth + 1)
+                        getThem(
```
Comment (ckcollab) [inline codalab/codalabtools/compute/worker.py]: Revert

File: codalab/codalabtools/compute/worker.py [hunk 5]
```diff
@@ -135,16 +145,33 @@ def getThem(bundle_id, bundle_rel_path, bundles, depth):
...
-                if k not in ("description", "command", "exitCode", "elapsedTime", "stdout", "stderr", "submitted-by", "submitted-at"):
+                if k not in (
+                        "description",
+                        "command",
+                        "exitCode",
+                        "elapsedTime",
+                        "stdout",
+                        "stderr",
+                        "submitted-by",
+                        "submitted-at"):
```
Comment (Raisins) [inline codalab/codalabtools/compute/worker.py]: :+1:

File: codalab/codalabtools/compute/tests.py [hunk 6]
```diff
@@ -6,19 +6,29 @@
...
 from codalabtools.compute.worker import WorkerConfig
 
+
 class ComputeConfigTests(TestCase):
+
     """Tests for WorkerConfig."""
 
     def no_config_test(self):
         """Loads non-existent config file."""
-        cfg_path = os.path.join(os.getcwd(), "codalabtools", "compute", "foo.config")
```
Comment (ckcollab) [inline codalab/codalabtools/compute/tests.py]: revert

File: codalab/codalabtools/compute/tests.py [hunk 7]
```diff
@@ -6,19 +6,29 @@
...
+            "codalabtools",
+            "compute",
+            "foo.config")
         self.assertFalse(os.path.exists(cfg_path))
         with self.assertRaises(EnvironmentError):
             WorkerConfig(cfg_path)
 
     def sample_config_test(self):
         """Loads 'sample.config' in codalabtools\\compute."""
-        cfg_path = os.path.join(os.getcwd(), "codalabtools", "compute", "sample.config")
```
Comment (ckcollab) [inline codalab/codalabtools/compute/tests.py]: revert

File: codalab/codalabtools/compute/worker.py [hunk 8]
```diff
@@ -73,9 +74,12 @@ def getAzureServiceBusQueue(self):
 
     def getLocalRoot(self):
         """Gets the path for the local directory where files are staged or None if the path is not provided."""
-        return self._winfo['local-root'] if 'local-root' in self._winfo else None
+        return self._winfo[
+            'local-root'] if 'local-root' in self._winfo else None
+
```
Comment (Raisins) [inline codalab/codalabtools/compute/worker.py]: 2nd line add's confusion.

File: codalab/apps/analytics/views.py (hunk 1 above)
Comment (ckcollab) [inline codalab/apps/analytics/views.py]: fixed

File: codalab/apps/api/routers.py (hunk 2 above)
Comment (ckcollab) [inline codalab/apps/api/routers.py]: fixed

File: codalab/codalabtools/compute/worker.py (hunk 4 above)
Comment (ckcollab) [inline codalab/codalabtools/compute/worker.py]: reverted

File: codalab/codalabtools/compute/tests.py (hunk 6 above)
Comment (ckcollab) [inline codalab/codalabtools/compute/tests.py]: reverted

File: codalab/codalabtools/compute/tests.py (hunk 7 above)
Comment (ckcollab) [inline codalab/codalabtools/compute/tests.py]: fixed

File: codalab/codalabtools/compute/worker.py (hunk 8 above)
Comment (ckcollab) [inline codalab/codalabtools/compute/worker.py]: fixed

**Comment by ckcollab** [comment]
@Raisins please test this vs worksheet stuff and let me know if it's working!

Also, @percyliang if you could give this a try I'd really appreciate it, tons of changes.

**Comment by percyliang** [comment]
This doesn't seem to work with the base64 changes from codalab-cli.
Perhaps need to merge in production-fixes or master?

Also, some of the reformatting is nice, but some of the splitting of
arguments onto multiple lines actually makes things harder to read - do we
really want to do that?

**Comment by ckcollab** [comment]
No we want clarity, please comment on the parts that should be reverted and I'll fix them up.

Can you write a failing test so I can develop against it? There should probably be a falling test anyway, maybe we can work on that together?

**Comment by ckcollab** [comment]
And thanks for testing it!

File: codalab/apps/api/tests.py [hunk 9]
```diff
@@ -158,10 +167,18 @@ def test_attempting_to_join_competition_auto_approved_sends_emails(self):
...
         self.assertIn('Accepted into Test Competition!', subjects)
-        self.assertIn('%s accepted into your competition!' % self.participant_user, subjects)
+        self.assertIn(
+            '%s accepted into your competition!' %
+            self.participant_user,
+            subjects)
 
-    def test_attempting_to_join_competition_not_logged_in_doesnt_send_email(self):
-        resp = self.client.post(reverse('competition-participate', kwargs={'pk': self.competition.pk}))
+    def test_attempting_to_join_competition_not_logged_in_doesnt_send_email(
```
Comment (percyliang) [inline codalab/apps/api/tests.py]: I noticed a lot of changes like these that spread out the lines, which doesn't actually seem to make it easier to read.  I guess this is because we have a 80 column width limit or something...

**Comment by percyliang** [comment]
Actually, the problem is not associated with this pull request, but a more
general problem, which I've filed an issue for here:
https://github.com/codalab/codalab-cli/issues/94
This PR looks good to me, although I agree we need more tests.

File: codalab/apps/api/tests.py [hunk 10]
```diff
@@ -206,9 +236,13 @@ def test_participation_status_update_revoked_sends_email(self):
...
         self.assertIn('Permission revoked from Test Competition!', subjects)
-        self.assertIn("%s's permission revoked from your competition!" % self.participant_user, subjects)
+        self.assertIn(
+            "%s's permission revoked from your competition!" %
+            self.participant_user,
+            subjects)
 
-    def test_participation_status_update_not_sent_when_participant_disables_status_notifications(self):
+    def test_participation_status_update_not_sent_when_participant_disables_status_notifications(
+            self):
```
Comment (Raisins) [inline codalab/apps/api/tests.py]: /n 
self):

is a bit much. prob better to put back on one line.

File: codalab/apps/api/tests.py [hunk 11]
```diff
@@ -261,7 +305,10 @@ def test_updating_participant_status_denied_without_competition_ownership(self):
     def test_updating_participant_status_works_as_owner(self):
         self.client.login(username="organizer", password="pass")
         resp = self.client.post(
-            reverse('competition-participation-status', kwargs={'pk': self.competition.pk}),
+            reverse(
+                'competition-participation-status',
+                kwargs={
+                    'pk': self.competition.pk}),
```
Comment (Raisins) [inline codalab/apps/api/tests.py]: easier to read on one line, same with above reverse.

File: codalab/apps/api/views/competition_views.py [hunk 12]
```diff
@@ -164,7 +158,8 @@ def publish(self, request, pk):
         c = webmodels.Competition.objects.get(id=pk)
         response = {}
         if self.request.user == c.creator or self.request.user in c.admins.all():
-            phases_needing_reference_data = webmodels.CompetitionPhase.objects.filter(competition=c, reference_data='').count()
+            phases_needing_reference_data = \
+                webmodels.CompetitionPhase.objects.filter(competition=c, reference_data='').count()
```
Comment (Raisins) [inline codalab/apps/api/views/competition_views.py]: back on one line was easier to understand and no `\` trickery

File: codalab/apps/api/views/worksheet_views.py [hunk 13]
```diff
@@ -6,42 +6,19 @@
...
-from django.contrib.sites.models import Site
-from django.core.exceptions import ObjectDoesNotExist
-from django.core.exceptions import PermissionDenied as DjangoPermissionDenied
-from django.core.files.base import ContentFile
-from django.core.mail import EmailMultiAlternatives
-from django.core.mail import send_mail
-from django.http import Http404, StreamingHttpResponse
-from django.template import Context
-from django.template.loader import render_to_string
-from django.utils.decorators import method_decorator
```
Comment (Raisins) [inline codalab/apps/api/views/worksheet_views.py]: Even though we are not using some of these inports bring back some of the more useful ones.  `PermissionDenied, ContentFile, EmailMultiAlternatives, send_mail, http imports` etc..

File: codalab/apps/authenz/forms.py [hunk 14]
```diff
@@ -30,4 +37,3 @@ def save(self, new_user):
             'bibtex': self.cleaned_data['bibtex'],
         })
         new_user.save()
-
```
Comment (Raisins) [inline codalab/apps/authenz/forms.py]: these changes make it harder to read as well.

File: codalab/apps/authenz/oauth.py [hunk 15]
```diff
@@ -20,12 +20,15 @@
...
+
 class Validator(OAuth2Validator):
+
     """
     Customizes the default validator.
     """
 
-    def validate_scopes(self, client_id, scopes, client, request, *args, **kwargs):
+    def validate_scopes(
+            self, client_id, scopes, client, request, *args, **kwargs):
```
Comment (Raisins) [inline codalab/apps/authenz/oauth.py]: no reason to be on 2 lines

File: codalab/apps/authenz/oauth.py [hunk 16]
```diff
@@ -78,8 +84,11 @@ def get_user_token(user):
         return None
 
     client = Application.objects.get(client_id=cli_client_id(user))
-    tokens = AccessToken.objects.filter(application_id=client.id,
-                                        expires__gt=timezone.now() + timedelta(minutes=5))
+    tokens = AccessToken.objects.filter(
+        application_id=client.id,
+        expires__gt=timezone.now() +
+        timedelta(
+            minutes=5))
```
Comment (Raisins) [inline codalab/apps/authenz/oauth.py]: revert.

File: codalab/apps/authenz/views.py [hunk 17]
```diff
@@ -106,10 +112,12 @@ def post(self, request, *args, **kwargs):
         token = d['token'] if 'token' in d else None
         scopes = d['scopes'] if 'scopes' in d else None
         response_data = ValidationApi._translate_token(token, scopes)
-        return HttpResponse(json.dumps(response_data), content_type="application/json")
+        return HttpResponse(
+            json.dumps(response_data), content_type="application/json")
```
Comment (Raisins) [inline codalab/apps/authenz/views.py]: this and the 3 pervious changes are a bit much. `"user").get(` does not help. lol 
better layout for above as well. 
```
'user': {
    'id': access_token.user.id, 
    'name': access_token.user.username
}
```

File: codalab/apps/forums/models.py [hunk 18]
```diff
@@ -25,7 +28,8 @@ def save(self, *args, **kwargs):
         return super(Thread, self).save(*args, **kwargs)
 
     def get_absolute_url(self):
-        return reverse('forum_thread_detail', kwargs={'forum_pk': self.forum.pk, 'thread_pk': self.pk })
+        return reverse('forum_thread_detail', kwargs={
+                       'forum_pk': self.forum.pk, 'thread_pk': self.pk})
```
Comment (Raisins) [inline codalab/apps/forums/models.py]: these changes reduce readability

File: codalab/apps/forums/urls.py [hunk 19]
```diff
@@ -4,8 +4,16 @@
...
+                       url(r'^(?P<forum_pk>\d+)/new_thread/$',
+                           views.CreateThreadView.as_view(),
+                           name='forum_new_thread'),
+                       url(r'^(?P<forum_pk>\d+)/(?P<thread_pk>\d+)/$',
+                           views.ThreadDetailView.as_view(),
+                           name='forum_thread_detail'),
+                       url(r'^(?P<forum_pk>\d+)/(?P<thread_pk>\d+)/new_post/$',
+                           views.CreatePostView.as_view(),
+                           name='forum_new_post'),
+                       )
```
Comment (Raisins) [inline codalab/apps/forums/urls.py]: you changed the prev urls but left these ones all broke up.

File: codalab/apps/forums/views.py [hunk 20]
```diff
@@ -86,4 +86,3 @@ class ThreadDetailView(ForumBaseMixin, DetailView):
     model = Thread
     template_name = "forums/thread_detail.html"
     pk_url_kwarg = 'thread_pk'
-
```
Comment (Raisins) [inline codalab/apps/forums/views.py]: these changes read better on one line.

File: codalab/apps/health/urls.py [hunk 21]
```diff
@@ -1,10 +1,15 @@
...
-    url(r'^check_thresholds', views.check_thresholds, name='health_status_check_thresholds')
-)
+                       url(r'^status', views.health, name='health_status'),
+                       url(r'^email_settings',
+                           views.email_settings,
+                           name='health_status_email_settings'),
+                       url(r'^check_thresholds',
+                           views.check_thresholds,
+                           name='health_status_check_thresholds')
+                       )
```
Comment (Raisins) [inline codalab/apps/health/urls.py]: see prev comment about urls.

File: codalab/apps/health/views.py [hunk 22]
```diff
@@ -17,16 +15,23 @@ def get_health_metrics():
     jobs_pending = Job.objects.filter(status=Job.PENDING)
     jobs_pending_count = len(jobs_pending)
 
-    jobs_finished_in_last_2_days = Job.objects.filter(status=Job.FINISHED, created__gt=datetime.now() - timedelta(days=2))
+    jobs_finished_in_last_2_days = Job.objects.filter(
+        status=Job.FINISHED,
+        created__gt=datetime.now() -
+        timedelta(
+            days=2))
```
Comment (Raisins) [inline codalab/apps/health/views.py]: The line break up makes it hard to follow the logic. Here is a proposed idea.  
```
jobs_finished_in_last_2_days = Job.objects.filter(
                                    status=Job.FINISHED, 
                                    created__gt=datetime.now() - timedelta(days=2)
                                )
```

File: codalab/apps/health/views.py [hunk 23]
```diff
@@ -34,7 +39,8 @@ def get_health_metrics():
         if (job.updated - job.created) > timedelta(minutes=10):
             jobs_lasting_longer_than_10_minutes.append(job)
 
-    jobs_failed = Job.objects.filter(status=Job.FAILED).order_by("-updated")[:10]
+    jobs_failed = Job.objects.filter(
+        status=Job.FAILED).order_by("-updated")[:10]
```
Comment (Raisins) [inline codalab/apps/health/views.py]: this the the prev 3 changes could used revered or updated no more `\` crazyness. lol line 29 while long made sense before the split

File: codalab/apps/health/views.py [hunk 24]
```diff
@@ -79,13 +84,19 @@ def check_thresholds(request):
...
 
-        if metrics["jobs_lasting_longer_than_10_minutes"] and len(metrics["jobs_lasting_longer_than_10_minutes"]) > 10:
-            send_mail("Codalab Warning: Many jobs taking > 10 minutes!", "There are many jobs taking longer than 10 minutes to process", settings.DEFAULT_FROM_EMAIL, emails)
+        if metrics["jobs_lasting_longer_than_10_minutes"] and len(
+                metrics["jobs_lasting_longer_than_10_minutes"]) > 10:
+            send_mail(
+                "Codalab Warning: Many jobs taking > 10 minutes!",
+                "There are many jobs taking longer than 10 minutes to process",
+                settings.DEFAULT_FROM_EMAIL,
+                emails)
```
Comment (Raisins) [inline codalab/apps/health/views.py]: if statement chagne ok,  send mail :+1:

**Comment by Raisins** [comment]
This seems to be fine code wise. Nothing is broken. Worksheets and other things work fine. 

Stylistically there is a lot. Many line and changes are for the worst in account of both reliability and code flow. Some lines that are better understood reading as a single line although long are now broken up and have logic jumps.with odd breaks around and in brackets and parentheses. 

pep8 is great and most of it's rules are helpful and if you're code is a travesty it's a great clean up but remember PEP 20 
 * Beautiful is better than ugly
 * Explicit is better than implicit
 * Simple is better than complex
 * Complex is better than complicated
 * Readability counts

I think too many of these changes were put in place to please some random pep8 error output. When in fact they where fine they where. Change pep8 requirement based on feeling, it's not a hard fast rule book it's suggestions.  
From the pep8 page
>A style guide is about consistency. Consistency with this style guide is important. Consistency within a project is more important. Consistency within one module or function is most important.

>But most importantly: know when to be inconsistent -- sometimes the style guide just doesn't apply. When in doubt, use your best judgment. Look at other examples and decide what looks best.

File: codalab/apps/api/tests.py [hunk 25]
```diff
@@ -242,13 +281,18 @@ def setUp(self):
         self.participant = CompetitionParticipant.objects.create(
             user=self.participant_user,
             competition=self.competition,
-            status=ParticipantStatus.objects.get(codename=ParticipantStatus.PENDING)
+            status=ParticipantStatus.objects.get(
+                codename=ParticipantStatus.PENDING)
         )
 
-    def test_updating_participant_status_denied_without_competition_ownership(self):
+    def test_updating_participant_status_denied_without_competition_ownership(
```
Comment (ckcollab) [inline codalab/apps/api/tests.py]: Please god no

File: codalab/apps/authenz/oauth.py [hunk 26]
```diff
@@ -90,7 +99,8 @@ def get_user_token(user):
         access_token = AccessToken(
             user=user,
             scope='',
-            expires=timezone.now() + timedelta(seconds=oauth2_settings.ACCESS_TOKEN_EXPIRE_SECONDS),
```
Comment (ckcollab) [inline codalab/apps/authenz/oauth.py]: fix
